w3af/plugins/grep/vulners_scanner.py

```diff
@@ -0,0 +1,219 @@
+
+    :author: Vulners.com Team: Kir Ermakov (isox@vulners.com)
+    """
+
+    def __init__(self):
+        GrepPlugin.__init__(self)
+
+        # Vulners shared objects
+        self._vulners_api = None
+        self._vulners_api_key = None
+        self.rules_table = None
+        self.rules_updated = False
+
+        self._already_visited = ScalableBloomFilter()
+        self._vulnerability_cache = {}
+
+    def get_options(self):
+        """
+        :return: A list of option objects for this plugin.
+        """
+        ol = OptionList()
+
+        d = ('Vulners API key for extended scanning rate limits.'
+             'You can obtain one for free at https://vulners.com')
+        o = opt_factory('vulners_api_key', self._vulners_api_key, d, STRING)
+        ol.add(o)
+        return ol
+
+    def set_options(self, options_list):
+        """
+        This method sets all the options that are configured using the user
+        interface generated by the framework using the result of get_options().
+        :param options_list: A dictionary with the options for the plugin.
+        :return: No value is returned.
+        """
+        self._vulners_api_key = options_list['vulners_api_key'].get_value()
+
+
+    def update_vulners_rules(self):
+        # Get fresh rules from Vulners.
+        try:
+            self.rules_table = self.get_vulners_api().rules()
+            # Adapt it for MultiRe structure [(regex,alias)] removing regex duplicated
+            regexAliases = collections.defaultdict(list)
+            for softwareName in self.rules_table:
+                regexAliases[self.rules_table[softwareName].get('regex')] += [softwareName]
+            # Now create fast RE filter
+            # Using re.IGNORECASE because w3af is modifying headers when making RAW dump. Why so? Raw must be raw!
+            self._multi_re = MultiRE(((regex, regexAliases.get(regex)) for regex in regexAliases), re.IGNORECASE)
+        except Exception as e:
+            self.rules_table = None
+            error_message = 'Vulners plugin failed to init with error: %s'
+            om.out.error(error_message % e)
+
+    def get_vulners_api(self):
+        # Lazy import. Just not to make it in the __init__
+        # Other way API will try to communicate Vulners right at the w3af startup
+        if not self._vulners_api:
+            # Don't forget to setup Vulners API key if you have one
+            self._vulners_api = vulners.Vulners(api_key= self._vulners_api_key)
+        return self._vulners_api
+
+    def check_vulners(self, software_name, software_version, check_type):
+        vulnerabilities = {}
+        cached_result = self._vulnerability_cache.get((software_name, software_version, check_type))
+        if cached_result:
+            return cached_result
+        # Ask Vulners about vulnerabilities
+        if check_type == 'software':
+            vulnerabilities = self.get_vulners_api().softwareVulnerabilities(software_name, software_version)
```

Comment:
What happens when the API triggers a 418 here?

Same two lines below in call to `cpeVulnerabilities`

Shouldn't we log at least the error so the user knows that he needs an API key?

Reply:
Also, when 418 is triggered for the first time, shouldn't we disable the grep plugin? If we hit the limit then it makes no sense to keep trying to match any regexes.

Reply:
API can't hit 418 when using Vulners Python API.
There is throttling inside. It reads response headers from the Vulners telling client API to slow down.